src/main/scala/io/iohk/ethereum/blockchain/sync/FastSync.scala

```diff
@@ -132,10 +132,10 @@ class FastSync(
         requestedHeaders.get(peer).foreach{ requestedNum =>
           removeRequestHandler(sender())
           requestedHeaders -= peer
-          if (blockHeaders.size <= requestedNum)
+          if (blockHeaders.nonEmpty && blockHeaders.size <= requestedNum && blockHeaders.head.number == syncState.bestBlockHeaderNumber + 1)
```

Comment:
Minor detail: What do you think of calling `checkHeadersChain` here instead of in `handleBlockHeaders`, so that all the validations are done in a single place? (or the other way around)

Reply:
Actually it is a good idea, but I would abstain from it for now as there is separate refactor ticket (https://iohk.myjetbrains.com/youtrack/issue/EC-524) and I would like to minimise code changes for now